README
Title: 0
Album: 33
Album: 63
Year: ?
Genre: ?

id3_client

   -offset mode: search and display data by start/end offset.
   -title mode: search and display by title
     ...


doc/CHANGELOG
2.9.1
------

-Defined offset in getTags as ID3_SIZE
-id3_client now takes two parameters to select fields


2.9.0
------
-absorb dirwalk into id3_file.c


2.8
----

-increased BUFF_SIZE to 8192
-fixed fclose on filename to close on filepointer


2.7
----

Implemented Directory Recursion (dirwalk and fsize);

defined BUFF_SIZE



2.6
----

-Assigned recordCount to 1st Byte of 1st Record.

Binding Variables:

char *filename      // The filename of an mp3       
char *dirPath       // The directory path to a directory tha contains mp3s.
char *filePath      // The path to an mp3
GList *filelist     // List of mp3s that are scanned
id3String *tagBuff  // The id3 Tags


rename checkFilename to isMp3
-change function name in getID3.h
-change function name in id3_file.c
-change function call in getFileList( char *) in id3_file.c

- get dirpath as argument in id3_server.c

-FullPath
  -should be in id3_data.c

return codes:

copyTags
getFileList


3.0
-----

mcast_sender.c - datagram support to transport data over UDP.


3.1
-----

-Objects and Source are renamed to achieve a better naming convention
-True multicast capability 


src/id3_data.c

/*
Author: Gregory Maglio
Date: 02/2011
Desc: module to handle transversale of allocated buffer of type id3Buff
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/ipc.h>
#include <sys/shm.h>
#include <dirent.h>
#include <glib.h>
#include <glib/glist.h>
#include "../headers/id3shm.h"

void printTags( id3Buff tagBuff ) {

   const int recordCount = tagBuff.bufSize;
   int i = 0;

   for( i = 0; i < recordCount;  i++) { 
        printf("%s\n", tagBuff.buf[i].id3String); 
   }
   printf("record count: %i\n", recordCount); 

}

id3String *copyTags( id3String *dest, id3String *src, const int size ) {

   int i = 0;

   for(i = 0; i < size; i++) {
        strncpy( dest[i].id3String, src[i].id3String, ID3_LENGTH );
   }
   return dest;

}

id3Buff *copyBuff( id3Buff *dest, id3Buff src ) {

   dest->bufSize = src.bufSize;
   copyTags(dest->buf, src.buf, src.bufSize);
   return dest;

}

id3String *getField( id3String *field, id3String tag, const unsigned int start, const unsigned int end ) {

   int i = 0;
   int j = 0;

   for( i = 0, j = start; j <= end; i++, j++ ) { field->id3String[i] = tag.id3String[j]; }
      
   return field; 

}


src/id3_file.c
/*
Author: Gregory Maglio
Date: 02/2011
Desc: module to manage file i/o to the end of populating the buffer with Payload.
*/


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/ipc.h>
#include <sys/shm.h>
#include <dirent.h>
#include <glib.h>
#include <glib/glist.h>
#include "../headers/id3shm.h"

int isMp3(char *filename) {

    char *extension = NULL;
    extension = strrchr(filename,'.');

    if( extension != NULL ) {  
        if( strcasecmp(extension, ".mp3") == 0 ) 
           return 0;
    }
    return 1;  

}

/* dirwalk: apply fcn to all files in dir */
void dirwalk(char *dir, GList **filelist, void (*fcn)(char *, GList **)) {

    char d_name[MAX_PATH];
    struct dirent *dp;
    DIR *dfd;
    char *fullpath = "\0";

    if ((dfd = opendir(dir)) == NULL) {
        fprintf(stderr, "dirwalk: can't open %s\n", dir);
        return;
    }

    while ((dp = readdir(dfd)) != NULL) {
        if (strcmp(dp->d_name, ".") == 0 || strcmp(dp->d_name, "..") == 0)
            continue;
                      /* skip self and parent */

        if (strlen(dir)+strlen(dp->d_name)+2 > sizeof(d_name))
            fprintf(stderr, "dirwalk: d_name %s %s too long\n", dir, dp->d_name);
        else {
            sprintf(d_name, "%s/%s", dir, dp->d_name);
            (*fcn)(d_name, filelist);
        }
    }
    closedir(dfd);
}

void getFileList(char *d_name, GList **filelist) {

    struct stat stbuf;
    char *fullpath = "\0";

    if (stat(d_name, &stbuf) == -1) {
        fprintf(stderr, "getFileList: can't access %s\n", d_name);
        // return;
    }
    if ((stbuf.st_mode & S_IFMT) == S_IFDIR)
        dirwalk(d_name, filelist, getFileList);
    if( isMp3(d_name) == 0 ) { 
        fullpath = calloc( FN_LENGTH, sizeof(char) );
        strcpy(fullpath, d_name);
        *filelist = g_list_prepend( *filelist, fullpath );
    }
}

int getTags( GList *filelist, id3String *tagBuff ) {

   FILE *fp;
   const int offset = -ID3_LENGTH;
   int i = 0;
 
   while(filelist != NULL) {
       if ( (fp = fopen(filelist->data,"r")) == NULL) {
          perror("Error: getTags(): file not opened");
          return -1;
       }
       fseek( fp, offset, SEEK_END );
       fgets( tagBuff[i].id3String, ID3_LENGTH, fp );
       i++;
       fclose(fp);
       filelist = g_list_next(filelist);

   }
   return i; 

}


src/id3_getshm.c
/*
Author: Gregory Maglio
Date: 02/2011
Desc: retrieve data from shared buffer
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/ipc.h>
#include <sys/shm.h>
#include <dirent.h>
#include <glib.h>
#include <glib/glist.h>
#include "../headers/id3shm.h"


void helper() {

    printf("id3_client <start offset> <end offset> <record index>\n");

}

int main(int argc, char *argv[]) {

   id3Buff *shmBuf;
   id3String title[TITLELENGTH];
   id3String album[ALBUMLENGTH];
   int recordCount = 0;
   int shmid = 0;
   int startOffset = 0;
   int endOffset = 0;
   int recordIndex = 0;
  
   if( (shmid = shmget( SHMKEY, sizeof(id3Buff), SHMPERM )) < 0 ) {
        perror("Failed to create Shared Memory Segment");
   }

   shmBuf = shmat( shmid, 0, 0);

   if(argc == 4) {
       startOffset = atoi(argv[1]);  
       endOffset = atoi(argv[2]);  
       recordIndex = atoi(argv[3]);  
   }
   else {
       printTags( *shmBuf );
       getTitle( title, shmBuf->buf[recordIndex] );
       printf("Title: %s\n", title->id3String);
       getAlbum( album, shmBuf->buf[recordIndex] );
       printf("Album: %s\n", album->id3String);
   }

   if( shmdt( shmBuf ) < 0 ) {
        perror("Failed to detach Shared Memory Segment");
   }
     
   return 0;

}


src/id3_putshm.c
/*
Author: Gregory Maglio
Date: 02/2011
Desc: Retrieve id3 info from files withing specified directory tree.  Copy id3 info to shared buffer.
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/ipc.h>
#include <sys/shm.h>
#include <dirent.h>
#include <glib.h>
#include <glib/glist.h>
#include "../headers/id3shm.h"


int main(int argc, char *argv[]) {

   id3Buff *shmBuff = NULL;
   id3Buff tagBuff;
   GList *filelist = NULL;
   char dirPath[FN_LENGTH] = "\0";
   int shmid = 0;
   int i = 0;

   if( argc < 2 ) {
        printf("Missing argument.\n"); 
        return -1;
   } 

   strcpy(dirPath,argv[1]);   // get argument from user
   getFileList(dirPath, &filelist); 
   tagBuff.bufSize = getTags(filelist, tagBuff.buf);
    
   printf("record count: %i\n", tagBuff.bufSize);
   if( tagBuff.bufSize > 1 ) { 
       if( (shmid = shmget( SHMKEY, sizeof(id3Buff), IPC_CREAT|IPC_EXCL|SHMPERM )) < 0 ) {
           perror("Failed to create Shared Memory Segment");
           return -1;
       }
       printf("shmget() is successful.\n");
   }
   else {
       printf("No records to process.\n");
       return tagBuff.bufSize;;
   }

   if( ( shmBuff = shmat( shmid, 0, 0)) < 0 ) {
        perror("Failed to attach to Shared Memory Segment");
   }
   copyBuff( shmBuff, tagBuff );
   if( shmdt( shmBuff ) < 0 ) {
        perror("Failed to detach Shared Memory Segment");
   }
   
   return 0;

}


src/id3_mcast_sender.c
/*
Author: Gregory Maglio
Date: 02/2011
Desc: Send id3 tag information through UDP Multicast.
*/



#include <sys/types.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/socket.h>
#include <netdb.h>
#include "../headers/id3shm.h"


int main(void) {

    id3Buff *shmBuf; // id3shm
    const int TTL = 64;
    struct sockaddr_in to;
    struct ip_mreq multiaddr;
    int shmid = 0; // id3shm
    int m = 0;
    int j = 0;
    int len = 0;

    int sock = socket(AF_INET, SOCK_DGRAM, 0);
   
 
    if( (shmid = shmget( SHMKEY, sizeof(id3Buff), SHMPERM )) < 0 ) { // id3shm
        perror("Failed to create Shared Memory Segment");           // id3shm
 	exit(EXIT_FAILURE);
    }                                                                // id3shm
    shmBuf = shmat( shmid, 0, 0);                                    //id3shm

    // Join a multicast group 224.1.1.1 on all interfaces
    multiaddr.imr_multiaddr.s_addr = inet_addr("224.1.1.1");
    multiaddr.imr_interface.s_addr = htonl(INADDR_ANY);
    setsockopt(sock, IPPROTO_IP, IP_ADD_MEMBERSHIP, &multiaddr, sizeof(multiaddr));
                     

    setsockopt(sock, IPPROTO_IP, IP_MULTICAST_TTL, &TTL, sizeof(TTL));
                     

    // setup the destination address of 224.1.1.1 and UDP port 2007
    to.sin_family = AF_INET;
    to.sin_addr.s_addr = inet_addr("224.1.1.1");
    to.sin_port = htons(2007);

    // write some data
    for (j = 3; j < shmBuf->bufSize; j++) {
        len = strlen(shmBuf->buf[j].id3String) + 1;
                /* +1 for terminating null byte */

        if (len + 1 > BUFF_SIZE) {
            fprintf(stderr,
                    "Ignoring long message in argument %d\n", j);
            continue;
        }

        if(sendto(sock, shmBuf->buf[j].id3String, len, 0, (const struct sockaddr*)&to,sizeof(to)) != len) {
            fprintf(stderr, "partial/failed write\n");
            exit(EXIT_FAILURE);
        }

    }
             

      // tell router we are no longer interested in this group
      close(sock);
      return 0;
}



src/id3_mcast_receiver.c
/*
Author: Gregory Maglio
Date: 02/2011
Desc: Listen for id3 tag information via UDP Multicast.
*/


#include <sys/types.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/socket.h>
#include <netdb.h>
#include <arpa/inet.h>
#include <netinet/in.h>
#include "../headers/id3shm.h"


int main(void) {

      char buffer[BUFF_SIZE];
      struct sockaddr_in to, buffer_sin;
      struct ip_mreq multiaddr;
      socklen_t addr_len;
      ssize_t numbytes;
      const int TTL = 64;
      int m;

      int sock; 
      if( (sock = socket(AF_INET, SOCK_DGRAM, 0)) < 0 ) {
             perror("socket");
             exit(1);
      }
      
      // setup UDP port 2007 and set address struct for any interface
      to.sin_family = AF_INET;
      to.sin_addr.s_addr = htonl(INADDR_ANY);
      to.sin_port = htons(2007);

      // bind
      if(bind(sock, (struct sockaddr *)&to, sizeof(to)) < 0) {
               perror("bind");
               exit(1);
       }

      // Join a multicast group 224.1.1.1 on all interfaces
      multiaddr.imr_multiaddr.s_addr = inet_addr("224.1.1.1");
      multiaddr.imr_interface.s_addr = htonl(INADDR_ANY);
      setsockopt(sock, IPPROTO_IP, IP_ADD_MEMBERSHIP, &multiaddr, sizeof(multiaddr));
                     

      addr_len = sizeof(buffer_sin);
      //read here
      while (1) {
            memset(buffer, '\0', sizeof(buffer));
            if((numbytes = recvfrom(sock, buffer, BUFF_SIZE-1, 0, (struct sockaddr *)&buffer_sin, &addr_len)) < 0) {
                
                    perror("recvfrom");
                    exit(1);
            }
            printf("%s\n", buffer); 
      }


      // tell router we are no longer interested in this group
      close(sock);
      return 0;
}


headers/id3shm.h
/* 
Gregory Maglio
Feburay 2011
Definition of ID3 structure and message
*/



#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/ipc.h>
#include <sys/shm.h>
#include <dirent.h>
#include <glib.h>
#include <glib/glist.h>

#define EXTENSION ".mp3"
#define PATH "/home/gmaglio"
#define ID3_LENGTH 128 // Length of ID3 Tags
#define FN_LENGTH 1024 // Length of filename
#define BUFF_SIZE 8192 // Length of filename
#define SHMKEY 0x1702
#define SHMPERM ( SHM_R|SHM_W|SHM_R>>3|SHM_W>>3|SHM_R>>6|SHM_W>>6 )
#define MAX_PATH 1024 // used by dirwalk()


#define TITLE 3 // Offset for Title
#define TITLELENGTH 29 // Length of Title
#define getTitle( field, tag ) getField( field, tag, TITLE, TITLE + TITLELENGTH )

#define ALBUM 33 // Offset for Album
#define ALBUMLENGTH 20 // Length of Album
#define getAlbum( field, tag ) getField( field, tag, ALBUM, ALBUM + ALBUMLENGTH )


typedef struct id3String {

     char id3String[ID3_LENGTH];

} id3String;


// derived from struct id3String
typedef struct id3Buff {
 
     id3String buf[BUFF_SIZE];
     int bufSize;

} id3Buff;


// returns 1 if not mp3
// used by getFileList
int isMp3( char *filename ); 

// id3_file
void getFileList(char *d_name, GList **filelist);
void dirwalk(char *, GList **, void (*fcn)(char *, GList **));
int getTags( GList *filelist, id3String *tagBuff );

// id3_data
void printTags( id3Buff tagBuff );
id3String *copyTags( id3String *dest, id3String *src, const int size );
id3Buff *copyBuff( id3Buff *dest, id3Buff src );
id3String *getField( id3String *field, id3String tag, const unsigned int start, const unsigned int end );
